### PR TITLE
lock stdout while writing json progress bar

### LINKED ISSUE #12225: --json progress is incorrect with parallel downloads
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

E.g. in conda create -n demo_0002 --json python=3.10 pandas command output I could see:
```
^@{"fetch":"pandas-1.5.2         | 11.9 MB   | ","finished":false,"maxval":1,"progress":0.079953}
^@{"fetch":"mkl-2021.4.0         | 142.6 MB  | ","finished":false,"maxval":1,"progress":0.009860}
^@{"fetch":"numpy-base-1.23.5    | 6.7 MB    | ","finished":false,"maxval":1,"progress":0.207858}
^@<94>h,<8C> c713f04495093c2e1a1e8590fc95633a<94>h0<8C>U[https://repo.anaconda.com/pkgs/main/linux-64/daal-devel{](https://repo.anaconda.com/pkgs/main/linux-64/daal-devel%7B)"fetch":"pandas-1.5.2         | 11.9 MB   | ","finished":false,"maxval":1,"progress":0.081264}
^@{"fetch":"mkl-2021.4.0         | 142.6 MB  | ","finished":false,"maxval":1,"progress":0.009970}
^@{"fetch":"pandas-1.5.2         | 11.9 MB   | ","finished":false,"maxval":1,"progress":0.082575}
^@{"fetch":"numpy-base-1.23.5    | 6.7 MB    | ","finished":false,"maxval":1,"progress":0.210194}
```

### Conda Info

`22.11.1`

### LINKED ISSUE #12231: Unexpected output with `--json` option in conda 22.11.*
### Checklist

- [X] I added a descriptive title
- [X] I searched open reports and couldn't find a duplicate

### What happened?

Starting with conda 22.11.0 (incl. 22.11.1) - running `conda` with `--json` flag may generate broken output. Example (part of the output of `conda create -n demo_0002 --json python=3.10 pandas`):
```
^@{"fetch":"pandas-1.5.2         | 11.9 MB   | ","finished":false,"maxval":1,"progress":0.079953}
^@{"fetch":"mkl-2021.4.0         | 142.6 MB  | ","finished":false,"maxval":1,"progress":0.009860}
^@{"fetch":"numpy-base-1.23.5    | 6.7 MB    | ","finished":false,"maxval":1,"progress":0.207858}
^@<94>h,<8C> c713f04495093c2e1a1e8590fc95633a<94>h0<8C>U[https://repo.anaconda.com/pkgs/main/linux-64/daal-devel{](https://repo.anaconda.com/pkgs/main/linux-64/daal-devel%7B)"fetch":"pandas-1.5.2         | 11.9 MB   | ","finished":false,"maxval":1,"progress":0.081264}
^@{"fetch":"mkl-2021.4.0         | 142.6 MB  | ","finished":false,"maxval":1,"progress":0.009970}
^@{"fetch":"pandas-1.5.2         | 11.9 MB   | ","finished":false,"maxval":1,"progress":0.082575}
^@{"fetch":"numpy-base-1.23.5    | 6.7 MB    | ","finished":false,"maxval":1,"progress":0.210194}
```
Where `^@` - is an expected delimiter between JSON records, and `<94>`, `<8C>` - control or non-unicode characters.
Line in the middle is not a valid JSON that breaks external tools in their attempts to parse it.

### Conda Info

```shell
active environment : None
            shell level : 0
       user config file : /home/user/.condarc
 populated config files : 
          conda version : 22.11.0
    conda-build version : not installed
         python version : 3.9.12.final.0
       virtual packages : __archspec=x86_64=0
                          __glibc=2.31=0
                          __linux=5.15.0=0
       base environment : /home/user/Development/Projects/Anaconda/.extra/conda_root  (writable)
      conda av data dir : /home/user/Development/Projects/Anaconda/.extra/conda_root/etc/conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/user/Development/Projects/Anaconda/.extra/conda_root/pkgs
                          /home/user/.conda/pkgs
       envs directories : /home/user/Development/Projects/Anaconda/.extra/conda_root/envs
                          /home/user/.conda/envs
               platform : linux-64
             user-agent : conda/22.11.0 requests/2.27.1 CPython/3.9.12 Linux/5.15.0-56-generic ubuntu/20.04.5 glibc/2.31
                UID:GID : 1001:1001
             netrc file : None
           offline mode : False
```

### Conda Config

```shell
<empty>
```

### Conda list

```shell
# packages in environment at /home/user/Development/Projects/Anaconda/.extra/conda_root:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main    defaults
_openmp_mutex             4.5                       1_gnu    defaults
brotlipy                  0.7.0           py39h27cfd23_1003    defaults
ca-certificates           2022.10.11           h06a4308_0    defaults
certifi                   2022.12.7        py39h06a4308_0    defaults
cffi                      1.15.0           py39hd667e15_1    defaults
charset-normalizer        2.0.4              pyhd3eb1b0_0    defaults
colorama                  0.4.4              pyhd3eb1b0_0    defaults
conda                     22.11.0          py39h06a4308_1    defaults
conda-content-trust       0.1.1              pyhd3eb1b0_0    defaults
conda-package-handling    1.8.1            py39h7f8727e_0    defaults
cryptography              36.0.0           py39h9ce1e76_0    defaults
idna                      3.3                pyhd3eb1b0_0    defaults
ld_impl_linux-64          2.35.1               h7274673_9    defaults
libffi                    3.3                  he6710b0_2    defaults
libgcc-ng                 9.3.0               h5101ec6_17    defaults
libgomp                   9.3.0               h5101ec6_17    defaults
libstdcxx-ng              9.3.0               hd4cf53a_17    defaults
ncurses                   6.3                  h7f8727e_2    defaults
openssl                   1.1.1s               h7f8727e_0    defaults
pip                       21.2.4           py39h06a4308_0    defaults
pluggy                    1.0.0            py39h06a4308_1    defaults
pycosat                   0.6.3            py39h27cfd23_0    defaults
pycparser                 2.21               pyhd3eb1b0_0    defaults
pyopenssl                 22.0.0             pyhd3eb1b0_0    defaults
pysocks                   1.7.1            py39h06a4308_0    defaults
python                    3.9.12               h12debd9_0    defaults
readline                  8.1.2                h7f8727e_1    defaults
requests                  2.27.1             pyhd3eb1b0_0    defaults
ruamel.yaml               0.16.12          py39h27cfd23_1    defaults
ruamel.yaml.clib          0.2.6            py39h7f8727e_0    defaults
ruamel_yaml               0.15.100         py39h27cfd23_0    defaults
setuptools                61.2.0           py39h06a4308_0    defaults
six                       1.16.0             pyhd3eb1b0_1    defaults
sqlite                    3.38.2               hc218d9a_0    defaults
tk                        8.6.11               h1ccaba5_0    defaults
toolz                     0.12.0           py39h06a4308_0    defaults
tqdm                      4.63.0             pyhd3eb1b0_0    defaults
tzdata                    2022a                hda174b7_0    defaults
urllib3                   1.26.8             pyhd3eb1b0_0    defaults
wheel                     0.37.1             pyhd3eb1b0_0    defaults
xz                        5.2.5                h7b6447c_0    defaults
yaml                      0.2.5                h7b6447c_0    defaults
zlib                      1.2.12               h7f8727e_1    defaults
```

### Additional Context

Similar behavior was also detected in other environments. Primarily these are GitHub Actions runners - `windows-latest`, `ubuntu-latest`, `macos-latest`, with different versions of python - from 3.7 and up to 3.10.

Issue happened in 90-95% of test runs in mentioned matrix of environments. What is common for all of them - each test run has its own fresh miniconda environment, which required downloading packages for `conda create` commands.

Locally, where I had everything cached - I didn't get similar issue. Unless I requested a packages that I didn't install previously in any environment.

### PR DESCRIPTION
**Pull request by dholth**
### Description

`sys.stdout` isn't threadsafe, leading to corrupted output. Lock while writing json progress bar substitute.

Fix #12225 
Fix #12231

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- [x] Add / update necessary tests?
- [x] Add / update outdated documentation?

### CODE REVIEW

File: conda/common/io.py [hunk 1]
```diff
@@ -1,32 +1,41 @@
...
+
+from conda._vendor.tqdm.std import TqdmDefaultWriteLock
+
 from ..auxlib.decorators import memoizemethod
 from ..auxlib.logz import NullHandler
 from ..auxlib.type_coercion import boolify
-from tqdm import tqdm
+from .compat import encode_environment, on_win
+from .constants import NULL
+from .path import expand
```
Comment (dholth) [inline conda/common/io.py:38]: sort imports

File: conda/common/io.py [hunk 2]
```diff
@@ -474,8 +491,11 @@ def __init__(self, description, enabled=True, json=False, position=None, leave=T
     def update_to(self, fraction):
         try:
             if self.json and self.enabled:
-                sys.stdout.write('{"fetch":"%s","finished":false,"maxval":1,"progress":%f}\n\0'
-                                 % (self.description, fraction))
+                with self.get_lock():
+                    sys.stdout.write(
+                        '{"fetch":"%s","finished":false,"maxval":1,"progress":%f}\n\0'
+                        % (self.description, fraction)
+                    )
```
Comment (dholth) [inline conda/common/io.py:498]: took the opportunity to format with black

File: conda/common/io.py [hunk 3]
```diff
@@ -433,7 +446,25 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+try:
+    from threading import RLock  # type: ignore
+except (ImportError, OSError):  # pragma: no cover
+    # according to tqdm, threading.RLock is not guaranteed
+    class RLock:
+        def __enter__(self):
+            pass
+
+        def __exit__(self, *exc):
+            pass
```
[the remark this answers is not in the document]
Comment (XuehaiPan) [reply conda/common/io.py:458]: `conda` now only supports Python 3.7+. We can guarantee `threading.RLock`.

**Comment by travishathaway** [comment]
@dholth, did you try to write a test for this to recreate the failing condition. The code itself looks reasonable, but it would be nice to have a new test case covering it.

I understand testing threading issues may be pretty difficult in this context though...

**Comment by dholth** [comment]
This is very easy to test manually, reliably corrupts output when piped into `tee`, correct output after locking added. We would need to pipe a parallel download to a file and assert no special characters - imo not a practical test.

File: conda/common/io.py (hunk 3 above)
Comment (jezdez) [inline conda/common/io.py:458]:
```suggestion
```

File: conda/common/io.py [hunk 4]
```diff
@@ -1,32 +1,39 @@
...
 from itertools import cycle
-import json
-import logging
-from logging import CRITICAL, Formatter, NOTSET, StreamHandler, WARN, getLogger
-import os
+from logging import CRITICAL, NOTSET, WARN, Formatter, StreamHandler, getLogger
 from os.path import dirname, isdir, isfile, join
-import signal
-from threading import Event, Thread, Lock
+from threading import Event, Lock, Thread
```
Comment (jezdez) [inline conda/common/io.py:26]:
```suggestion
from threading import Event, Lock, RLock, Thread
```

**Review by jezdez — CHANGES_REQUESTED** [review]
This looks ready to go, except for the comment by @XuehaiPan about RLock being available in conda's 3.7+ world.

@dholth Could you elaborate about where tqdm says `threading.RLock` isn't always available?

**Comment by dholth** [comment]
> 

https://github.com/tqdm/tqdm/blob/master/tqdm/std.py#L69

**Comment by dholth** [comment]
I removed the RLock fallback since we require Python 3.7+

**Review by jezdez — APPROVED** [review]

**Pull request merged**